### ISSUE TITLE
error: 'PTHREAD_MUTEX_RECURSIVE' undeclared (first use in this function)

### ISSUE THREAD

**Issue by dverma05** [openssl/openssl#14092]
Getting below error:
  ```
crypto/threads_pthread.c: In function 'CRYPTO_THREAD_lock_new':
crypto/threads_pthread.c:48:38: error: 'PTHREAD_MUTEX_RECURSIVE' undeclared (first use in this function)
crypto/threads_pthread.c:48:38: note: each undeclared identifier is reported only once for each function it appears in
make[1]: *** [crypto/threads_pthread.o] Error 1

 ```
Below is the configdata.pm dump:
```

Command line (with current working directory = .):

    /usr/bin/perl ./Configure shared --prefix=/home/dileep/PROJECT_NAME/opensource/target/abc_temp no-zlib no-zlib-dynamic linux-generic32 no-async

Perl information:

    /usr/bin/perl
    5.18.2 for i686-linux-gnu-thread-multi-64int

Enabled features:

    afalgeng
    aria
    asm
    autoalginit
    autoerrinit
    autoload-config
    bf
    blake2
    buildtest-c\+\+
    camellia
    capieng
    cast
    chacha
    cmac
    cms
    comp
    ct
    deprecated
    des
    dgram
    dh
    dsa
    dso
    dtls
    dynamic-engine
    ec
    ec2m
    ecdh
    ecdsa
    engine
    err
    filenames
    gost
    hw(-.+)?
    idea
    makedepend
    md4
    mdc2
    multiblock
    nextprotoneg
    pinshared
    ocb
    ocsp
    pic
    poly1305
    posix-io
    psk
    rc2
    rc4
    rdrand
    rfc3779
    rmd160
    scrypt
    seed
    shared
    siphash
    sm2
    sm3
    sm4
    sock
    srp
    srtp
    sse2
    ssl
    static-engine
    stdio
    tests
    threads
    tls
    ts
    ui-console
    whirlpool
    tls1
    tls1-method
    tls1_1
    tls1_1-method
    tls1_2
    tls1_2-method
    tls1_3
    dtls1
    dtls1-method
    dtls1_2
    dtls1_2-method

Disabled features:

    asan                    [default] OPENSSL_NO_ASAN
    async                   [option]  OPENSSL_NO_ASYNC
    crypto-mdebug           [default] OPENSSL_NO_CRYPTO_MDEBUG
    crypto-mdebug-backtrace [default] OPENSSL_NO_CRYPTO_MDEBUG_BACKTRACE
    devcryptoeng            [default] OPENSSL_NO_DEVCRYPTOENG
    ec_nistp_64_gcc_128     [default] OPENSSL_NO_EC_NISTP_64_GCC_128
    egd                     [default] OPENSSL_NO_EGD
    external-tests          [default] OPENSSL_NO_EXTERNAL_TESTS
    fuzz-libfuzzer          [default] OPENSSL_NO_FUZZ_LIBFUZZER
    fuzz-afl                [default] OPENSSL_NO_FUZZ_AFL
    heartbeats              [default] OPENSSL_NO_HEARTBEATS
    md2                     [default] OPENSSL_NO_MD2 (skip crypto/md2)
    msan                    [default] OPENSSL_NO_MSAN
    rc5                     [default] OPENSSL_NO_RC5 (skip crypto/rc5)
    sctp                    [default] OPENSSL_NO_SCTP
    ssl-trace               [default] OPENSSL_NO_SSL_TRACE
    ubsan                   [default] OPENSSL_NO_UBSAN
    unit-test               [default] OPENSSL_NO_UNIT_TEST
    weak-ssl-ciphers        [default] OPENSSL_NO_WEAK_SSL_CIPHERS
    zlib                    [option]  
    zlib-dynamic            [option]  
    ssl3                    [default] OPENSSL_NO_SSL3
    ssl3-method             [default] OPENSSL_NO_SSL3_METHOD

Config target attributes:

    AR => "ar",
    ARFLAGS => "r",
    CC => "gcc",
    CFLAGS => "-Wall -O3",
    CXX => "g++",
    CXXFLAGS => "-Wall -O3",
    HASHBANGPERL => "/usr/bin/env perl",
    RANLIB => "ranlib",
    RC => "windres",
    aes_asm_src => "aes_core.c aes_cbc.c",
    aes_obj => "aes_core.o aes_cbc.o",
    apps_aux_src => "",
    apps_init_src => "",
    apps_obj => "",
    bf_asm_src => "bf_enc.c",
    bf_obj => "bf_enc.o",
    bn_asm_src => "bn_asm.c",
    bn_obj => "bn_asm.o",
    bn_ops => "BN_LLONG RC4_CHAR",
    build_file => "Makefile",
    build_scheme => [ "unified", "unix" ],
    cast_asm_src => "c_enc.c",
    cast_obj => "c_enc.o",
    cflags => "-pthread",
    chacha_asm_src => "chacha_enc.c",
    chacha_obj => "chacha_enc.o",
    cmll_asm_src => "camellia.c cmll_misc.c cmll_cbc.c",
    cmll_obj => "camellia.o cmll_misc.o cmll_cbc.o",
    cppflags => "",
    cpuid_asm_src => "mem_clr.c",
    cpuid_obj => "mem_clr.o",
    cxxflags => "-std=c++11 -pthread",
    defines => [  ],
    des_asm_src => "des_enc.c fcrypt_b.c",
    des_obj => "des_enc.o fcrypt_b.o",
    disable => [  ],
    dso_extension => ".so",
    dso_scheme => "dlfcn",
    ec_asm_src => "",
    ec_obj => "",
    enable => [ "afalgeng" ],
    ex_libs => "-ldl -pthread",
    exe_extension => "",
    includes => [  ],
    keccak1600_asm_src => "keccak1600.c",
    keccak1600_obj => "keccak1600.o",
    lflags => "",
    lib_cflags => "",
    lib_cppflags => "-DOPENSSL_USE_NODELETE",
    lib_defines => [  ],
    md5_asm_src => "",
    md5_obj => "",
    modes_asm_src => "",
    modes_obj => "",
    module_cflags => "-fPIC",
    module_cxxflags => "",
    module_ldflags => "-Wl,-znodelete -shared -Wl,-Bsymbolic",
    padlock_asm_src => "",
    padlock_obj => "",
    poly1305_asm_src => "",
    poly1305_obj => "",
    rc4_asm_src => "rc4_enc.c rc4_skey.c",
    rc4_obj => "rc4_enc.o rc4_skey.o",
    rc5_asm_src => "rc5_enc.c",
    rc5_obj => "rc5_enc.o",
    rmd160_asm_src => "",
    rmd160_obj => "",
    shared_cflag => "-fPIC",
    shared_defflag => "-Wl,--version-script=",
    shared_defines => [  ],
    shared_extension => ".so.\$(SHLIB_VERSION_NUMBER)",
    shared_extension_simple => ".so",
    shared_ldflag => "-Wl,-znodelete -shared -Wl,-Bsymbolic",
    shared_rcflag => "",
    shared_sonameflag => "-Wl,-soname=",
    shared_target => "linux-shared",
    thread_defines => [  ],
    thread_scheme => "pthreads",
    unistd => "<unistd.h>",
    uplink_aux_src => "",
    uplink_obj => "",
    wp_asm_src => "wp_block.c",
    wp_obj => "wp_block.o",

Recorded environment:

    AR = i486-openwrt-linux-ar
    ARFLAGS = 
    AS = i486-openwrt-linux-as
    ASFLAGS = 
    BUILDFILE = 
    CC = i486-openwrt-linux-gcc
    CFLAGS = -Os
    CPP = i486-openwrt-linux-cpp
    CPPDEFINES = 
    CPPFLAGS = -I/home/dileep/PROJECT_NAME/opensource/target/noarch/include -I/home/dileep/PROJECT_NAME/opensource/../resources/toolchains/abc_temp/include
    CPPINCLUDES = 
    CROSS_COMPILE = 
    CXX = 
    CXXFLAGS = 
    HASHBANGPERL = 
    LD = i486-openwrt-linux-ld
    LDFLAGS = -L/home/dileep/PROJECT_NAME/opensource/target/abc_temp/lib -L/home/dileep/PROJECT_NAME/opensource/../resources/toolchains/abc_temp/lib
    LDLIBS = 
    MT = 
    MTFLAGS = 
    OPENSSL_LOCAL_CONFIG_DIR = 
    PERL = 
    RANLIB = i486-openwrt-linux-ranlib
    RC = 
    RCFLAGS = 
    RM = 
    WINDRES = 
    __CNF_CFLAGS = 
    __CNF_CPPDEFINES = 
    __CNF_CPPFLAGS = 
    __CNF_CPPINCLUDES = 
    __CNF_CXXFLAGS = 
    __CNF_LDFLAGS = 
    __CNF_LDLIBS = 

Makevars:

    AR              = i486-openwrt-linux-ar
    ARFLAGS         = r
    AS              = i486-openwrt-linux-as
    CC              = i486-openwrt-linux-gcc
    CFLAGS          = -Os
    CPP             = i486-openwrt-linux-cpp
    CPPDEFINES      = 
    CPPFLAGS        = -I/home/dileep/PROJECT_NAME/opensource/target/noarch/include -I/home/dileep/PROJECT_NAME/opensource/../resources/toolchains/abc_temp/include
    CPPINCLUDES     = 
    CXX             = g++
    CXXFLAGS        = -Wall -O3
    HASHBANGPERL    = /usr/bin/env perl
    LD              = i486-openwrt-linux-ld
    LDFLAGS         = -L/home/dileep/PROJECT_NAME/opensource/target/abc_temp/lib -L/home/dileep/PROJECT_NAME/opensource/../resources/toolchains/abc_temp/lib
    LDLIBS          = 
    PERL            = /usr/bin/perl
    RANLIB          = i486-openwrt-linux-ranlib
    RC              = windres
    RCFLAGS         = 

NOTE: These variables only represent the configuration view.  The build file
template may have processed these variables further, please have a look at the
build file for more exact data:
    Makefile

build file:

    Makefile

build file templates:

    Configurations/common0.tmpl
    Configurations/unix-Makefile.tmpl
    Configurations/common.tmpl
dileep@dileep-VirtualBox:~/PROJECT_NAME/opensource/openssl/openssl-1.1.1i$

**Comment by dverma05** [comment]
Configuring -D_XOPEN_SOURCE=700 is working for me but still would like to bring you notice. if you want to close this issue you are open to close.

**Comment by mattcaswell** [comment]
What platform is this?

**Comment by dverma05** [comment]
Ubuntu 14.4
toolchain: crosstools-arm-gcc-5.3-linux-4.1-glibc-2.22-binutils-2.25

**Comment by mahender08** [comment]
@t8m do we have any update on this ? 

I am also facing same issue :- https://github.com/openssl/openssl/issues/19449

@dverma05 how did you resolve this problem ?

**Comment by levitte** [comment]
> Configuring -D_XOPEN_SOURCE=700 is working for me but still would like to bring you notice. if you want to close this issue you are open to close.

... which would indicate a requirement for [POSIX 2008](https://pubs.opengroup.org/onlinepubs/9699919799.2008edition/).  I thought that was normally enable by default (since quite a long time ago), but apparently not with the `i486-openwrt-linux` cross compilation.

**Comment by levitte** [comment]
> @dverma05 how did you resolve this problem ?

They added `-D_XOPEN_SOURCE=700` on the configuration command line.

**Comment by mahender08** [comment]
@levitte , thanks let me try

**Comment by mahender08** [comment]
@levitte  , no luck

 command : **./Configure enable-fips D_XOPEN_SOURCE=700 --prefix=/usr/local/ssl --openssldir=/usr/local/ssl**

![image](https://user-images.githubusercontent.com/88478603/196903158-71afc5de-793c-477f-8477-fc9da38880d6.png)

**Comment by t8m** [comment]
Closing in favor of #19449
